Add documentation for orographic gravity wave (#1677)
CliMA/ClimaAtmos.jl: docs/src/gravity_wave.md

  X(z_{n-1}) = 0.5*\left[X(z_{n-3/2}) +X(z_{n-1/2}) \right].
  ```
By applying the above parameterization on zonal and meridional winds, the forcing for the physical wind velocity is computed. We further transform them onto the Covariant vectors and that will be the tendencied added onto the momentum equation.

## Orographic gravity wave
The orographic gravity wave drag parameterization follows the methods described in [garner2005](@cite). The momentum drag from sub-grid scale mountains is divided into a non-propagating component and a propagating component. The non-propagating component forces momentum drag within the planetary boundary layer while the propagating component generate a `c=0` gravity wave which propagates upwards and deposit momentum flux to the layers where it breaks.

### PBL top

> There are many ways to determine the PBL top. We implement the following simple criteria to find the PBL top level k as the hight level that satisfies

Reviewer: ```suggestion
There are many ways to determine the PBL top. We implement the following simple criteria to find the PBL top level k as the highest level that satisfies
```

```math
^c p[k] \ge 0.5 ∗ ^f p[0]
```
and
```math
^c T[0] + T_{boost} - ^c T[k] > g/c_p * ( ^c z[k] - ^c z[0] )
```

> where the superscripts ``f`` and ``c`` represent cell faces and cell centers in the vertical stencils. ``T_{boost} = 1.5`` is the surface temperature boost to improve PBL height estimate.

Reviewer: ```suggestion
where the superscripts ``f`` and ``c`` represent cell faces and cell centers in the vertical stencils. ``T_{boost} = 1.5 \mathrm{K}`` is the surface temperature boost to improve PBL height estimate.
```

### Orographic information

> The orographic information needed in generating the base momentum flux for low-level flow encountering the sub-grid scale mountains. We compute the tensor ``\textbf{T}`` and the scalar ``h_{max}`` from the Earth elevation data (codes [here](https://caltech.box.com/s/w4szffattzofarpwyv9rmm5s77jgo3o4))。 They describe all relevant information including amplitude, variance, orientation, and anisotropy about topography.

Reviewer: ```suggestion
The orographic information needed in generating the base momentum flux for low-level flow encountering the sub-grid scale mountains. We compute the tensor ``\textbf{T}`` and the scalar ``h_{max}`` from the Earth elevation data (codes [here](https://caltech.box.com/s/w4szffattzofarpwyv9rmm5s77jgo3o4)). They describe all relevant information including amplitude, variance, orientation, and anisotropy about topography.
```
😀

### Base flux
The base momentum flux generated is computed and divided into the propagating and non-propagating components.

Let ``\overline{\cdot}`` represents the mean property of the low-level flow which can be obtained

> as either the average within PBL or the value at the first cell center right above PBL top. Let ``\overline{V} = (\overline{u}, \overline{v})``, ``\overline{N}``, and ``\overline(\rho)`` represent the horizontal wind, buoyancy frequency, and density of the low-level flow. ``\overline{N}`` is computed as 

Reviewer: ```suggestion
as either the average within PBL or the value at the first cell center right above PBL top. Let ``\overline{V} = (\overline{u}, \overline{v})``, ``\overline{N}``, and ``\overline{\rho}`` represent the horizontal wind, buoyancy frequency, and density of the low-level flow. ``\overline{N}`` is computed as 
```

```
```math
FrU_{max} = \max(Fr_{max} * U_{sat}, FrU_{min} + \epsilon_0),
```
```math
FrU_{clp} = \min(FrU_{max}, \max(FrU_{min}, FrU_{sat})),
```
```math
FrU_0 = \frac{U_0}{V_{\tau}}U_{sat},
```
and ``U_0=1.0`` is the arbitrary velocity scale.

Now the correct linear drag is computed as 
```math
\tau_l = \frac{FrU_{max}^{2+\gamma-\epsilon} - FrU_{min}^{2+\gamma-\epsilon}}{2+\gamma-\epsilon},
```
and the propagating and non-propagating parts of the drag are computed as
```math
\tau_p = a_0 \left[ \frac{FrU_{clp}^{2+\gamma-\epsilon} - FrU_{min}^{2+\gamma-\epsilon}}{2+\gamma-\epsilon} + FrU_{sat}^{\beta+2} \frac{FrU_{max}^{\gamma-\epsilon-\beta} - FrU_{clp}^{\gamma-\epsilon-\beta}}{\gamma-\epsilon-\beta} \right],
```
```math
\tau_{np} = a_1 \frac{U_{sat}}{1+\beta} \left[ \frac{FrU_{max}^{1+\gamma-\epsilon} - FrU_{clp}^{1+\gamma-\epsilon}}{1+\gamma-\epsilon} - FrU_{sat}^{\beta+1} \frac{FrU_{\max}^{\gamma-\epsilon-\beta} - FrU_{clp}^{\gamma-\epsilon-\beta}}{\gamma-\epsilon-\beta} \right].
```

Here, $(\gamma, \epsilon, \beta) = (0.4, 0.0, 0.5)$ are arbitrary parameters that describe the mountain shapes.

### Saturation profiles for the propagating component

> The vertical profiles of saturated momentum flux ``\tau_{sat}`` is computed then so that momentum forcing can be obtained for 11d\overline{\arrow{V}}/dt = -\overline{\rho}^{-1}d\tau_{sat}/dz11. This only applies to the propagating part.

Reviewer: ```suggestion
The vertical profiles of saturated momentum flux ``\tau_{sat}`` is computed then so that momentum forcing can be obtained for ``d\overline{\arrow{V}}/dt = -\overline{\rho}^{-1}d\tau_{sat}/dz``. This only applies to the propagating part.
```

Similar to the base flux calculation but for the 3D fields, we computed ``N`` and ``V_\tau`` at cell faces as
```math

> ^f N[k]^2 = \frac{g}{^f T[k]} * ( ^f \overline{\frac{dT}{dz}}[k] + g/cp),

Reviewer: ```suggestion
^f N[k]^2 = \frac{g}{^f T[k]} * ( ^f \overline{\frac{dT}{dz}}[k] + \frac{g}{cp}),
```

```
```math
^f V_{\tau}[k] = \max(\epsilon_0, - V[k] \cdot \frac{\tau}{|\tau|}).
```

Let ``L_1 = L_0 * \max(0.5, \min(2.0, 1.0-samp*V_\tau*d^2V_{\tau}/N^2))`` where ``samp=1.0`` is the correction for coarse sampling of ``d^2V/dz^2``, and
```math
^f d^2V_{\tau}[k] = - \frac{d^2 V}{dz^2}[k] \cdot \frac{\tau}{|\tau|}.
```
The saturated velocity ``U_{sat}`` is refined as follows and used to computed the intermediate ``FrU's``
```math
U_{sat} = \min(U_{sat}, \sqrt{\frac{^\mathrm{f} \rho}{\rho_0} \frac{^\mathrm{f} V_{\pmb{\tau}}^3}{^\mathrm{f}N L_1} }).
```

> The ``FrU_{min}`` and ``FrU_{`max}`` are inherited from the base flux calculation. Let's save the source level ``FrU_{sat}`` and ``FrU_{clp}`` into

Reviewer: ```suggestion
The ``FrU_{min}`` and ``FrU_{max}`` are inherited from the base flux calculation. Let's save the source level ``FrU_{sat}`` and ``FrU_{clp}`` into
```

```

For level ``k`` between PBL top and ``kref``, the forcing due to non-propagating component is
```math
^c \left( \frac{du}{dt} [k] \right)_{np} = g \frac{\tau_{np}}{\tau_l} \frac{weight[k]}{wtsum} \tau_x, 
```
```math
^c \left( \frac{dv}{dt} [k] \right)_{np} = g \frac{\tau_{np}}{\tau_l} \frac{weight[k]}{wtsum} \tau_y, 
```
where ``(\tau_x, \tau_y, \tau_l, \tau_{np})`` is computed in the base flux calculation.

### Constrain the forcings
Total drag from both components are
```math
^c \left( \frac{du}{dt} [k] \right)_{\tau} = ^c \left( \frac{du}{dt} [k] \right)_{p} + ^c \left( \frac{du}{dt} [k] \right)_{np},
```
```math
^c \left( \frac{dv}{dt} [k] \right)_{\tau} = ^c \left( \frac{dv}{dt} [k] \right)_{p} + ^c \left( \frac{dv}{dt} [k] \right)_{np}.
```
To avoid instability due to large tendencies from the forcing, let's constrain the forcing magnitude with ``\epsilon_V = 3e-3``, and let
```math
^c \left( \frac{du}{dt} [k] \right)_{\tau} = \max(-\epsilon_V, \min(\epsilon_V, ^c \left( \frac{du}{dt} [k] \right)_{\pmb{\tau}})),
```
```math
^c \left( \frac{dv}{dt} [k] \right)_{\tau} = \max(-\epsilon_V, \min(\epsilon_V, ^c \left( \frac{dv}{dt} [k] \right)_{\pmb{\tau}})).
```

> Here we computed the forcing on the physical velocity (i.e., zonal and meridional wind). They are converted to the Covariant12Vecctor before added to ``Y_t`` in the codes.

Reviewer: ```suggestion
Here we computed the forcing on the physical velocity (i.e., zonal and meridional wind). They are converted to the Covariant12Vector before being added to ``Y_t`` in the codes.
```
